horovod/common/controller.cc

```diff
@@ -225,6 +242,19 @@ ResponseList Controller::ComputeResponseList(std::atomic_bool& shut_down) {
         }
       }
 
+      // Check if tensors from previous ticks are ready to reduce after Joins.
+      if (state.joined_size > 0) {
+        for (auto& table_iter : message_table_) {
+          int count = (int)table_iter.second.size();
+          if (count == (size_ - state.joined_size) &&
+              std::find(ready_to_reduce.begin(), ready_to_reduce.end(),
```

Comment:
Can you explain why this scan through the list is necessary?  Under what circumstances would the tensor already be in the `ready_to_reduce` list?

Reply:
The code scans for tensors that are currently **not** in the `ready_to_reduce` list, but have enough `table_iter.second.size()` that now, after new rank joined, they need to be in `ready_to_reduce`.

Reply:
I understand that, but my question is: why is it necessary to check if the tensor is in `ready_to_reduce` before doing the insertion?  Why not just insert it without the check?  I am not sure how it's possible that a tensor could already be in the list at the time this condition is met.  Can multiple `table_iter` objects meet the `count == (size_ - state.joined_size` condition for the same `table_iter.first` object?

Reply:
I see. I think you're right, no need to check. I'll fix it.

Reply:
I've tested this, and the check is actually needed because about 40 lines before this the coordinator can add to  `ready_to_reduce` based in its own messages.